serde/src/de/buffer.rs

```diff
@@ -0,0 +1,1806 @@
+//! Data structures for buffering self-describing formats.
+//!
+//! ```
+//! # use serde::de::{Deserialize, value, IntoDeserializer, buffer::Buffer};
+//! # let source = value::U64Deserializer::<value::Error>::new(32);
+//! // Deserialize any self-describing format from `source` into `buffer`.
+//! let buffer = Buffer::deserialize(source).unwrap();
+//! // Turn the buffer back into a deserializer.
+//! let deserializer = IntoDeserializer::<value::Error>::into_deserializer(buffer);
+//! # assert_eq!(u32::deserialize(deserializer).unwrap(), 32);
+//! ```
+
+use lib::*;
+
+use __private::size_hint;
+
+use de::{self, Deserialize, Deserializer, EnumAccess, Expected, MapAccess, SeqAccess, Visitor};
+
+/// An efficient buffer for self-describing formats.
+#[derive(Debug, Clone)]
+pub struct Buffer<'de>(pub(crate) BufferInner<'de>);
+
+impl<'de> Buffer<'de> {
+    /// Tries to extract a string from the buffer.
+    ///
+    // TODO: This function is used from within code generated by `serde_derive` and, thus,
+    // needs to be public. If this is undesired, we can could simply have it as a normal
+    // function in `private::de` instead.
+    pub fn as_str(&self) -> Option<&str> {
+        match self.0 {
+            BufferInner::Str(x) => Some(x),
+            BufferInner::String(ref x) => Some(x),
+            BufferInner::Bytes(x) => str::from_utf8(x).ok(),
+            BufferInner::ByteBuf(ref x) => str::from_utf8(x).ok(),
+            _ => None,
+        }
+    }
```

Comment:
This function is used by code generated by `serde_derive`. How should we deal with that?